src/storage.rs

```diff
@@ -0,0 +1,175 @@
+use anyhow::{Error, bail};
+use aws_sdk_s3::error::SdkError;
+use aws_sdk_s3::operation::get_object::GetObjectError;
+use aws_sdk_s3::operation::head_object::HeadObjectError;
+use aws_sdk_s3::primitives::ByteStream;
+use reqwest::StatusCode;
+use std::path::Path;
+use tokio::runtime::Runtime;
+
+pub(crate) enum Storage {
+    ReadOnly(CdnReader),
+    ReadWrite(S3Storage),
+}
+
+impl Storage {
+    pub(crate) fn file_status(&self, path: &str) -> Result<FileStatus, Error> {
+        if let Some(hash) = self.get_file(&format!("{path}.sha256"))? {
+            Ok(FileStatus::Present {
```

Comment:
Should we also check that the actual file also exists?

Reply:
Is there a scenario you have in mind where the `.sha256` file could be present but the actual file be missing?

The `.sha256` is written after the whole file is uploaded, so in theory there should be no case where using the tool results in a stray `.sha256`. The only thing that comes to mind is if someone manually goes and deletes files from the bucket.

Reply:
Yeah, I just meant a potential inconsistency where the file got deleted through other means. But we don't need to worry about that now, I suppose.